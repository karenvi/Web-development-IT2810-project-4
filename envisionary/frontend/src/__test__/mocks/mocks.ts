
import { ADD_REVIEW } from "../../graphql/mutations";
import { GET_COUNTRIES, GET_REVIEWS_BY_COUNTRY_NAME } from "../../graphql/queries";





export const mocks = [
    {
      request: {
        query: GET_COUNTRIES,
        variables: {}
      },
      result: {
        data: {
            country: 
            [{ _id: "1", Area: "120538", CCA3: "PRK", Capital: "Pyongyang", Continent: "Asia", Country: "North Korea", Density: "216.2755", 
            GrowthRate: "1.0038", Population1970: "14996879", Population1980: "17973650", Population1990: "20799523", Population2000: "23367059",
            Population2010: "24686435", Population2015: "25258015", Population2020: "25867467", Population2022: "26069416", Rank: "56", 
            WorldPopulationPercentage: "0.33" },

            { _id: "2", Area: "25713", CCA3: "MKD", Capital: "Skopje", Continent: "Europe", Country: "North Macedonia", Density: "81.4218", GrowthRate: "0.9954",
            Population1970: "1656783", Population1980: "1907023", Population1990: "2044174", Population2000: "2037936", Population2010: "2093828",
            Population2015: "2107962", Population2020: "2111072", Population2022: "2093599", Rank: "150", WorldPopulationPercentage: "0.03"},

            { _id: "3", Area: "464", CCA3: "NFK", Capital: "Saipan", Continent: "Oceania", Country: "Northern Mariana Islands", Density: "106.7909", 
            GrowthRate: "1.0014", Population1970: "10143", Population1980: "17613", Population1990: "48002", Population2000: "80338", Population2010: "54087",
            Population2015: "51514", Population2020: "49587", Population2022: "49551", Rank: "210", WorldPopulationPercentage: "0"},
            ]
        }
      }
    },
    {
      request: {
        query: GET_REVIEWS_BY_COUNTRY_NAME,
        variables: { country: "Afghanistan" }
      },
      result: { 
        data: {
          countryByName: {
            Reviews: [
              { Name: "Trond", ReviewText: "Flott land 책 feriere i." , Date: "2022-10-26T13:50:35.783Z", Rating: 3.5 },
              { Name: "Hans og Grethe", ReviewText: "Fint og fredelig land for et idyllisk opphold p책 all-inclusive.", Date: "2022-10-26T13:53:34.973Z",  Rating: 5 },
              { Name: "Lisbeth", ReviewText: "Synes det var litt skummelt i Afghanistan.", Date: "2022-10-26T19:51:35.080Z", Rating: 3 }
            ]
          }
        }
      }  
    },
    {
      request: {
        query: ADD_REVIEW,
        variables: { country: "Afghanistan", reviewText: "Lorem ipsum dolor sit amet", date: "2022-10-26T13:53:34.973Z", rating: 3.5 }
      },
      restult: {
        data: {
          countryByName: {
            Reviews: [
              { Name: "Trond", ReviewText: "Flott land 책 feriere i." , Date: "2022-10-26T13:50:35.783Z", Rating: 3.5 },
              { Name: "Hans og Grethe", ReviewText: "Fint og fredelig land for et idyllisk opphold p책 all-inclusive.", Date: "2022-10-26T13:53:34.973Z",  Rating: 5 },
              { Name: "Lisbeth", ReviewText: "Synes det var litt skummelt i Afghanistan.", Date: "2022-10-26T19:51:35.080Z", Rating: 3 }
            ]
          }
        }
      }
    }
  ];


export const failMock = [
  {
    request: {
        query: GET_REVIEWS_BY_COUNTRY_NAME,
        variables: { country: "Afghanistan" }
    },
    error: new Error()
  }
];

export const noReviewsMock = [
  {
    request: {
      query: GET_REVIEWS_BY_COUNTRY_NAME,
      variables: { country: "Afghanistan" }
    },
    result: { 
      data: {
        countryByName: { Reviews: null }
      }
    }  
  }
];