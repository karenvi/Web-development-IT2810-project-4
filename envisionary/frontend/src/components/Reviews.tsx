import { useLocation } from 'react-router-dom';
import { useQuery } from '@apollo/client';
import { GET_COUNTRY_DATA_BY_NAME } from '../graphql/queries';
import { IReview } from "../types"
import PaginationReviews from "./PaginationReviews";
import { textStyling } from '../styles/StyleObjects';
import { useContext } from 'react';
import { ThemeContext } from '../App';

function Reviews() {
  const { theme } = useContext(ThemeContext);
  const location = useLocation();
  const { loading, error, data } = useQuery(GET_COUNTRY_DATA_BY_NAME, { variables: { country: location.state.country.Country } });
  
  // Theme
  const textStyle = {...(theme === "light" ? textStyling.light : textStyling.dark),};

  if (loading) return <p style={textStyle}>Loading reviews ...</p>;
  if (error) return <p style={textStyle}>Could not get reviews</p>;

  // Order reviews from new-old
  const sortReviews = () => {
    if (data.countryByName.Reviews) {
      let reviews = [...data?.countryByName?.Reviews];
      reviews.sort((firstReview: IReview, nextReview: IReview) => Date.parse(nextReview.Date) - Date.parse(firstReview.Date));
      return reviews
    }
    return [];
  }

  return (
    <>
      <PaginationReviews sortReviews={sortReviews()} country={location.state.country.Country} />
    </>
  );
}
export default Reviews;