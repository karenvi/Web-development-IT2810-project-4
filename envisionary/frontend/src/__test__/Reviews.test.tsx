import { MockedProvider } from '@apollo/client/testing';
import renderer from 'react-test-renderer';
import { failMock, mocks } from './mocks/mocks';
import { render, screen } from '@testing-library/react';
import Reviews from '../components/Reviews';
export {};

/**
 * Mocking useLocation in order to make the query mocks work for useQuery in component.
 * Technique found at https://blog.changani.me/2022/01/30/webdev/react/react-mocking-react-router-dom-s-use-location-with-jest/
 */
jest.mock('react-router-dom', () => {
    return {
        ...jest.requireActual('react-router-dom'),
        useLocation: () => {
            return {
                pathname: "/country",
                search: "",
                hash: "",
                state: {
                    country: {
                        Country: "Afghanistan",
                    }
                },
                key: "eb77n1u7"
            }
        }
    };
});

describe("Testing Reviews component", () => {
    it("snapshot test", () => {

        const component: renderer.ReactTestRenderer = renderer.create(
                <MockedProvider mocks={mocks} addTypename={false}>
                    <Reviews />
                </MockedProvider>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot(); 
    });

    it('Loads reviews correctly', async () => {
        render(
            <MockedProvider mocks={mocks} addTypename={false}>
                    <Reviews />
            </MockedProvider>
        );

        expect(await screen.findByText('Loading reviews ...')).toBeInTheDocument();
        // Trond
        expect(await screen.findByText('Trond')).toBeInTheDocument();
        expect(await screen.findByText('Flott land å feriere i.')).toBeInTheDocument();
        // Hans og Grethe
        expect(await screen.findByText('Hans og Grethe')).toBeInTheDocument();
        expect(await screen.findByText('Fint og fredelig land for et idyllisk opphold på all-inclusive.')).toBeInTheDocument();
        // Lisbeth
        expect(await screen.findByText('Lisbeth')).toBeInTheDocument();
        expect(await screen.findByText('Synes det var litt skummelt i Afghanistan.')).toBeInTheDocument();
        
    });

    it('Should show error message', async () => {
        render(
            <MockedProvider mocks={failMock} addTypename={false}>
                    <Reviews />
            </MockedProvider>
        );

        expect(await screen.findByText("Could not get reviews")).toBeInTheDocument();
    });

});