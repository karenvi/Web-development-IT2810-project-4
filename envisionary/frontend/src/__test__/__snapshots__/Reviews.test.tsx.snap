// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Reviews component snapshot test 1`] = `
<p>
  Loading reviews ...
</p>
`;
